cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -74,36 +84,56 @@ setupLocalStateQueryExpr ::
      -- Protocols must wait until 'waitDone' returns because premature exit will
      -- cause other incomplete protocols to abort which may lead to deadlock.
   -> Maybe ChainPoint
-  -> TMVar (Either Net.Query.AcquireFailure a)
-  -> LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO a
+  -> TMVar (Either (QueryError e) a)
+  -> NodeToClientVersion
+  -> ExceptT (QueryError e) (LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO) a
   -> Net.Query.LocalStateQueryClient (BlockInMode mode) ChainPoint (QueryInMode mode) IO ()
-setupLocalStateQueryExpr waitDone mPointVar' resultVar' f =
+setupLocalStateQueryExpr waitDone mPointVar' resultVar' ntcVersion f =
   LocalStateQueryClient . pure . Net.Query.SendMsgAcquire mPointVar' $
     Net.Query.ClientStAcquiring
-    { Net.Query.recvMsgAcquired = runContT (runLocalStateQueryExpr f) $ \result -> do
-        atomically $ putTMVar resultVar' (Right result)
+    { Net.Query.recvMsgAcquired = runContT (runReaderT (runLocalStateQueryExpr (runExceptT f)) ntcVersion) $ \result -> do
+        atomically $ putTMVar resultVar' result
         void $ atomically waitDone -- Wait for all protocols to complete before exiting.
         pure $ Net.Query.SendMsgRelease $ pure $ Net.Query.SendMsgDone ()
 
     , Net.Query.recvMsgFailure = \failure -> do
-        atomically $ putTMVar resultVar' (Left failure)
+        atomically $ putTMVar resultVar' (Left (QueryErrorAcquireFailure failure))
         void $ atomically waitDone -- Wait for all protocols to complete before exiting.
         pure $ Net.Query.SendMsgDone ()
     }
 
--- | Use 'queryExpr' in a do block to construct monadic local state queries.
-queryExpr :: QueryInMode mode a -> LocalStateQueryExpr block point (QueryInMode mode) r IO a
-queryExpr q =
-  LocalStateQueryExpr . ContT $ \f -> pure $
-    Net.Query.SendMsgQuery q $
-      Net.Query.ClientStQuerying
-      { Net.Query.recvMsgResult = f
-      }
+-- | Get the node server's Node-to-Client version.
+getNtcVersion :: ExceptT (QueryError e) (LocalStateQueryExpr block point (QueryInMode mode) r IO) NodeToClientVersion
+getNtcVersion =  lift $ LocalStateQueryExpr $ ReaderT pure
+
+-- | Lift a query value into a monadic query expression.
+-- Use 'queryExpr' in a do block to construct monadic local state queries.
+queryExpr :: QueryInMode mode a -> ExceptT (QueryError e) (LocalStateQueryExpr block point (QueryInMode mode) r IO) a
```

Comment:
I disagree with lifting this into `ExceptT` because then this is now imposed on users. Also we don't seem to gain much by doing this outside of things looking a little nicer in the cli.

Reply:
So if I understand your comment, it's okay to use `ExceptT` in CLI, but not API?

Reply:
This commit removes `ExceptT` from type signatures of exported functions in `Cardano.Api.IPC.Monad`. https://github.com/input-output-hk/cardano-node/pull/3358/commits/c51800580d7c924a52c4af06c984fb47b7f5a45d